arangod/Scheduler/SocketTask.cpp

```diff
@@ -230,32 +230,37 @@ void SocketTask::addToReadBuffer(char const* data, std::size_t len) {
 
 // does not need lock
 void SocketTask::resetKeepAlive() {
-  if (_useKeepAliveTimer) {
-    asio_ns::error_code err;
-    _keepAliveTimer->expires_from_now(_keepAliveTimeout, err);
-    if (err) {
-      closeStream();
-      return;
-    }
-
-    _keepAliveTimerActive.store(true, std::memory_order_relaxed);
-    auto self = shared_from_this();
-    _keepAliveTimer->async_wait([self, this](const asio_ns::error_code& error) {
-      if (!error) {  // error will be true if timer was canceled
-        LOG_TOPIC(ERR, Logger::COMMUNICATION)
-            << "keep alive timout - closing stream!";
-        closeStream();
-      }
-    });
+  if (!_useKeepAliveTimer) {
+    return;
   }
+  
+  // expires_from_now cancels pending operations
+  asio_ns::error_code err;
+  _keepAliveTimer->expires_from_now(_keepAliveTimeout, err);
+  if (err) {
+    closeStream();
+    return;
+  }
+  _keepAliveTimerActive.store(true);
+  
+  std::weak_ptr<arangodb::rest::Task> self = shared_from_this();
+  _keepAliveTimer->async_wait([self, this](const asio_ns::error_code& ec) {
+    if (!ec) {  // error will be true if timer was canceled
+      LOG_TOPIC(INFO, Logger::COMMUNICATION)
+      << "keep alive timout - closing stream!";
+      if (auto s = self.lock()) {
+        this->closeStream();
```

Comment:
I would change `self` to be of type `std::weak_ptr<SocketTask>`. Then you don't have to capture `this` and could simply use `s` instead. I would prefer this, because we may only access `this` after we successfully locked `self`.

Reply:
I know but `arangodb::rest::Task` is a virtual base class so static_cast does not work

Reply:
Right, `shared_from_this` returns a `shared_ptr<Task>`. But you can explicitly cast it:
```
std::weak_ptr<SocketTask> self = std::static_pointer_cast<SocketTask>(shared_from_this());
```